### ISSUE TITLE
Dense Diff Pool with Sparse Adjacency Matrix

### ISSUE THREAD

**Issue by danjenson** [pyg-team/pytorch_geometric#7364]
### 🚀 The feature, motivation and pitch

We have a GAT-based model where the first embed/pool layer is a `GATConv` and the remaining after `dense_diff_pool` are `DenseGATConv`. Unfortunately, `dense_diff_pool` seems to require a dense adjacency matrix. I have tried to use a sparse {COO, CSR, CSC} adjacency matrix, but they are not properly strided.

```python
    def forward(self, x, edge_index, batch):
        x = self.embed[0](x, edge_index)
        s = self.pool[0](x, edge_index)
        x, mask = to_dense_batch(x, batch)
        s, _ = to_dense_batch(s, batch)
        # TODO(danj): memory blowout
        adj = to_dense_adj(edge_index, batch)
        x, adj, link_loss, entropy_loss = dense_diff_pool(x, adj, s, mask)
        for i in range(1, self.n_steps):
            x = self.embed[i](x, adj)
            s = self.pool[i](x, adj)
            x, adj, ll, el = dense_diff_pool(x, adj, s)
            link_loss += ll
            entropy_loss += el
        return x.squeeze(), link_loss, entropy_loss
```

Is there any way to get this to execute without calling `adj = to_dense_adj(edge_index, batch)`?

**Comment by rusty1s** [comment]
Currently no, `dense_diff_pool` is designed to operate on dense adjacency matrices :( I don't think we can make it work for sparse adjacency matrices, but let me investigate.

**Comment by pi314ever** [comment]
Looking into the implementation for `dense_diff_pool`, I don't think it's possible to utilize sparseness of the adjacency matrix to reduce memory usage due to the linked loss $|| A - ss^T||_2$. As far as I know, this cannot be computed efficiently as $ss^T$ is dense.
